src/coreclr/jit/compiler.h

```diff
@@ -3538,6 +3538,9 @@ class Compiler
                                    GenTree*               op2,
                                    CorInfoInlineTypeCheck typeCheckInliningResult);
     GenTree* gtFoldExprCall(GenTreeCall* call);
+#ifdef FEATURE_HW_INTRINSICS
+    GenTree* gtFoldHWIntrinsic(GenTreeHWIntrinsic* node);
```

Comment:
Nit: the `gtFold*` family of functions are to be called from `gtFoldExpr`.

For the optimizing functions called from morph, there is a different pattern - [`fgOptimize*`](https://github.com/dotnet/runtime/blob/7ed266eba93f82bdb8fc41e9b7e8ed9685f6128f/src/coreclr/jit/compiler.h#L6399-L6421). It would be nice to follow that pattern.

Reply:
I can look at the other pattern `fgOptimize` - though, I think having these HW intrinsic optimizations should be next to the call optimizations such as `Enum.HasFlag/etc.`.

Reply:
The `gtFold*` functions are supposed to be callable "from any phase" (I am sure some aren't, because there is no great consistence in the compiler), notably from the importer, while the `fgOptimize*` functions are only to be called from morph (and can e. g. use local assertions or call other "morph-only" functions, or be reasoned about as "a pre-order function" and "post-order function"), so that's the primary reason I had this suggestion. If you feel the optimizations here (current and future) better fit with the contract of `gtFold*`, then let it be how it is.

Reply:
I see, that makes sense to me. The intent of this is really only for morph, so it probably makes more sense to put it under the `fgOptimize*` functions.